File: les/ext/coin/osi_sym_solver_interface.cc
// -*- coding: utf-8; mode: c++; -*-
//
// Copyright (c) 2013 Oleksandr Sviridenko
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <coin/CoinPackedVector.hpp>
#include <coin/OsiSymSolverInterface.hpp>

#include <boost/python.hpp>

using namespace boost::python;

struct OsiSymSolverInterfaceWrap : OsiSymSolverInterface,
                                   wrapper<OsiSymSolverInterface> {
public:
  OsiSymSolverInterfaceWrap() : OsiSymSolverInterface() {}

  virtual void addCol(const CoinPackedVector& vec, const double collb,
                      const double colub, const double obj) {
    OsiSymSolverInterface::addCol(vec, collb, colub, obj);
  }

  virtual void addRow(const CoinPackedVectorBase& vec,
                      const char rowsen, const double rowrhs,
                      const double rowrng) {
    OsiSymSolverInterface::addRow(vec, rowsen, rowrhs, rowrng);
  }

  virtual void setInteger(int i) {
    OsiSymSolverInterface::setInteger(i);
  }

  virtual bool setSymParam(const std::string key, int value) {
    return OsiSymSolverInterface::setSymParam(key, value);
  }

  virtual boost::python::object default_getColSolution() const {
    const double* solution = OsiSymSolverInterface::getColSolution();
    boost::python::list lst;
    for (int i = 0; i < OsiSymSolverInterface::getNumCols(); i++) {
      lst.append(solution[i]);
    }
    return lst;
  }
};

BOOST_PYTHON_MODULE(_osi_sym_solver_interface)
{
  class_<OsiSymSolverInterfaceWrap, boost::noncopyable>("OsiSymSolverInterface")
    // Solve methods
    .def("branch_and_bound", &OsiSymSolverInterfaceWrap::branchAndBound)
    .def("initial_solve", &OsiSymSolverInterfaceWrap::initialSolve)
    .def("resolve", &OsiSymSolverInterfaceWrap::resolve)
    .def("multi_criteria_branch_and_bound",
         &OsiSymSolverInterfaceWrap::multiCriteriaBranchAndBound)
    // Methods returning info on how the solution process terminated
    .def("is_abandoned", &OsiSymSolverInterfaceWrap::isAbandoned)
    .def("is_proven_optimal", &OsiSymSolverInterfaceWrap::isProvenOptimal)
    .def("is_proven_primal_infeasible",
         &OsiSymSolverInterfaceWrap::isProvenPrimalInfeasible)
    .def("is_proven_dual_infeasible",
         &OsiSymSolverInterfaceWrap::isProvenDualInfeasible)
    .def("is_iteration_limit_reached",
         &OsiSymSolverInterfaceWrap::isIterationLimitReached)
    .def("is_time_limit_reached", &OsiSymSolverInterfaceWrap::isTimeLimitReached)
    .def("is_target_gap_reached", &OsiSymSolverInterfaceWrap::isTargetGapReached)
    // Methods to expand a problem
    .def("add_col", &OsiSymSolverInterfaceWrap::addCol)
    .def("add_row", &OsiSymSolverInterfaceWrap::addRow)
    // Problem query methods
    .def("get_num_cols", &OsiSymSolverInterfaceWrap::getNumCols)
    .def("get_num_rows", &OsiSymSolverInterfaceWrap::getNumRows)
    .def("get_num_elements", &OsiSymSolverInterfaceWrap::getNumElements)
    .def("get_obj_sense", &OsiSymSolverInterfaceWrap::getObjSense)
    .def("get_infinity", &OsiSymSolverInterfaceWrap::getInfinity)
    // Methods to set variable type
    .def("set_integer", &OsiSymSolverInterfaceWrap::setInteger)
    // Solution query methods
    .def("get_col_solution", &OsiSymSolverInterfaceWrap::default_getColSolution)
    .def("get_obj_value", &OsiSymSolverInterfaceWrap::getObjValue)
    // Methods to modify the objective, bounds, and solution
    .def("set_obj_sense", &OsiSymSolverInterfaceWrap::setObjSense)
    // Parameter set/get methods
    .def("set_sym_param", &OsiSymSolverInterfaceWrap::setSymParam)
    ;
}

File: les/ext/coin/osi_sym_solver_interface.py
#!/usr/bin/env python
#
# -*- coding: utf-8; -*-
#
# Copyright (c) 2012-2013 Oleksandr Sviridenko
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import itertools

from les.ext.coin import _osi_sym_solver_interface
from les.ext.coin import coin_utils
from les.problems.problem_base import ProblemBase
from les.solvers.bilp_solver_base import BILPSolverBase

MINIMIZATION = +1
MAXIMIZATION = -1

class OsiSymSolverInterface(_osi_sym_solver_interface.OsiSymSolverInterface,
                            BILPSolverBase):
  """This class represents OSI Solver Interface for SYMPHONY."""

  def __init__(self):
    BILPSolverBase.__init__(self)
    _osi_sym_solver_interface.OsiSymSolverInterface.__init__(self)
    self._problem = None

  def solve(self):
    """"Invokes solver's built-in enumeration algorithm - branch and bound."""
    self.branch_and_bound()

  def load_problem(self, problem, details={}):
    """Loads problem to the solver.

    TODO: use OsiSymSolverInterface.loadProblem().
    """
    if not isinstance(problem, ProblemBase):
      raise TypeError("Problem must be derived from ProblemBase: %s" % type(problem))
    if not self._problem:
      details = {}
    # Setup objective functions
    if details.get("obj_coefs", True):
      # XXX: At some point we can not set column type when there is only one
      # column in the problem. Otherwise SYMPHONY crashes with segmentation
      # fault.
      if problem.get_num_cols() > 1:
        for i, coef in enumerate(problem.get_obj_coefs()):
          (p, v) = coef
          col = coin_utils.CoinPackedVector()
          # NOTE: fix coef because of C++ signature
          self.add_col(col, 0., 1., v.astype(float))
          # TODO: set column type
          self.set_integer(i)
      else:
        col = coin_utils.CoinPackedVector()
        self.add_col(col, 0., 1., problem.get_obj_coefs().values()[0].astype(float))
        self.set_integer(1)
      # Set objective function sense
      self.set_obj_sense(MAXIMIZATION)
    # Setup constraints
    if details.get("cons_matrix", True):
      for p, row in enumerate(problem.get_cons_matrix()):
        if not row.getnnz():
          continue
        r = coin_utils.CoinPackedVector();
        for i, v in itertools.izip(row.indices, row.data):
          r.insert(int(i), v.astype(float))
        # NOTE: fix coef because of C++ signature
        self.add_row(r, "L", float(problem.get_rhs()[p]), 1.)
    elif details.get("rhs", True):
      for i in xrange(len(problem.get_rhs())):
        self.set_row_upper(i, problem.get_rhs()[i].astype(float))
    self._problem = problem
